packages/passport-interface/src/CredentialManager.ts

```diff
@@ -0,0 +1,145 @@
+
+  // Can we get a credential containing a given PCD type?
+  public canGenerateCredential(req: CredentialRequest): boolean {
+    if (req.pcdType === "email-pcd") {
+      return this.pcds.getPCDsByType(req.pcdType).length !== 0;
+    } else if (req.pcdType === undefined) {
+      return true;
+    } else {
+      // We can't generate credentials containing any other PCD type yet
+      return false;
+    }
+  }
+
+  // Get a credential from the local cache, if it exists
+  private getCachedCredential(type?: string): SerializedPCD | null {
+    const cacheKey = type ?? "none";
+    const entry = this.cache.get(cacheKey);
+    if (entry) {
+      if (Date.now() - entry.timestamp < CREDENTIAL_CACHE_MAX_AGE) {
+        return entry.value;
+      } else {
+        this.cache.delete(cacheKey);
+      }
+    }
+
+    return null;
+  }
+
+  private setCachedCredential(
+    type: string | undefined,
+    value: SerializedPCD
+  ): void {
+    const cacheKey = type ?? "none";
+
+    this.cache.set(cacheKey, { value, timestamp: Date.now() });
+  }
+
+  /**
+   * Generates the requested credential, if possible.
+   * Takes a {@link CredentialRequest} and produces a serialized PCD which
+   * consists of a signature PCD (e.g. a semaphore signature PCD) which wraps
+   * a {@link FeedCredentialPayload}. This payload contains a timestamp, and
+   * may contain a PCD if a) the feed requests one and b) CredentialManager
+   * can find a matching PCD.
+   */
+  public async requestCredential(
+    req: CredentialRequest
+  ): Promise<SerializedPCD> {
+    const cachedCredential = this.getCachedCredential(req.pcdType);
+    if (cachedCredential) {
+      return cachedCredential;
+    }
+
+    // This is currently the only supported PCD for credential embedding
+    if (req.pcdType === "email-pcd") {
+      const cachedCredential = this.getCachedCredential(req.pcdType);
+      if (cachedCredential) {
+        return cachedCredential;
+      }
+
+      const pcds = this.pcds.getPCDsByType(req.pcdType);
+      if (pcds.length === 0) {
+        throw new Error(
+          `Could not find a PCD of type ${req.pcdType} for credential payload`
+        );
+      }
+      // In future we might want to support multiple email PCDs, but this
+      // works for now
+      const pcd = pcds[0];
+      const serializedPCD = await this.pcds.serialize(pcd);
+      const result = await this.signPayload(
+        createFeedCredentialPayload(serializedPCD)
+      );
+      this.cache.set(req.pcdType, { value: result, timestamp: Date.now() });
+      return result;
+    } else if (req.pcdType === undefined) {
+      const result = await this.signPayload(createFeedCredentialPayload());
+      this.setCachedCredential(req.pcdType, result);
+      return result;
+    } else {
+      throw new Error(
+        `Cannot issue credential containing a PCD of type ${req.pcdType}`
+      );
+    }
+  }
+
+  // Takes a payload and wraps it in a signature PCD.
+  private async signPayload(
```

Comment:
unless you intended for this to be able to sign with other signature schemes, maybe worth calling this `semaphoreSignPayload`